platforms/core-configuration/core-serialization-codecs/src/main/kotlin/org/gradle/internal/serialize/codecs/core/WorkNodeCodec.kt

```diff
@@ -55,81 +76,319 @@ typealias IdForNode = (Node) -> Int
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        val result = IdentityHashMap<ActionNode, List<Node>>()
+        partialResultsRef.get().forEach { nodeAndPostExecutionNodes ->
+            nodeAndPostExecutionNodes.forEach { (node, postExecutionNodes) ->
+                result[node] = postExecutionNodes
+            }
+        }
+        return result
+    }
+
+
+    private
+    fun ReadContext.readNodes(): NodeForId {
+        val nodeIdCount = readSmallInt()
+        val partialResultsRef = AtomicReference<PersistentList<List<NodeId>>>(PersistentList.of())
+        val projectProvider = getSingletonProperty<ProjectProvider>()
+        val groupPaths = readCollectionInto<Path, MutableList<Path>>(::ArrayList) {
+            Path.path(readString())
+        }
+
+        runBuildOperations(parallel = parallelLoad, message = "reading state nodes") {
+            groupPaths.map { groupPath ->
+                OperationInfo(displayName = "Loading $groupPath", progressDisplayName = groupPath.path) {
+                    contextSource.readContextFor(this@readNodes, groupPath).readWith(Unit) {
+                        setSingletonProperty(projectProvider)
+                        val partialResult = readGroupedNodes()
+                        partialResultsRef.updateAndGet {
+                            it.plus(partialResult)
+                        }
+                    }
+                }
+            }
+        }
+        val nodesById = Array<Node?>(nodeIdCount) { null }
+        partialResultsRef.get().forEach { ids ->
+            ids.forEach { (node, id) ->
+                nodesById[id] = node
+            }
+        }
+        return { id: Int -> nodesById[id]!! }
+    }
+
+    private
+    fun <R> IsolateContext.unwrapBuildOperationExceptions(message: String, action: () -> R): R =
+        try {
+            action()
+        } catch (e: MultipleBuildOperationFailures) {
+            val cause = when (val first = e.causes.first()) {
+                is BuildOperationInvocationException -> first.cause!!
+                else -> first
+            }
+            e.causes.subList(1, e.causes.size).forEach(cause::addSuppressed)
+
+            when (cause) {
+                is GradleException -> throw cause
+                is Exception -> {
+                    onError(cause) {
+                        text("Error while $message")
+                    }
+                    // never runs
+                    error(Unit)
+                }
+                else -> throw e
+            }
+        }
+
+    data class NodeId(
+        val node: Node,
+        val id: Int
+    )
+
+    private
+    fun NodeOwner.path(): Path {
+        // TODO:parallel-cc make sure there are no conflicts between project node and owner-less nodes
+        return when (this) {
+            NodeOwner.None -> owner.identityPath.child("owner-less")
```

Comment:
❓ Will this conflict with a project named `owner-less`?

Reply:
Using relative (with a fixed prefix for build identity) paths vs absolute project identity paths, and having a file name generation strategy differentiating between relative and absolute paths.